### PR TITLE
server: require web login if an env var is set

### PR DESCRIPTION
**Pull request by vilterp**
This PR introduces the temporary environment variable `COCKROACH_EXPERIMENTAL_REQUIRE_WEB_LOGIN ` to enable testing login functionality before a fully functional login/logout UI is merged. Supersedes #24944, which hardcoded this.

Once we have a functional login/logout UI, we will probably hardcode this variable to true, so that login will always be required when the cluster is in secure mode.

Before this change:
- Admin server endpoints don't care whether there is a logged-in user or not, even in secure mode; they always run as the root user (#8767)
- There is no way to log in — the `/_auth/v1/login` endpoint 404s

After this change:
- secure mode
    - with env var `COCKROACH_EXPERIMENTAL_REQUIRE_WEB_LOGIN=true`:
        - Admin server endpoints require _some user_ to be logged in, although they still run queries (e.g. against `crdb_internal`) as the root user.
        - `/_auth/v1/login` is exposed, allowing clients such as the admin UI to log in.
    - else:
        - no auth (same as before)
- insecure mode
    - no auth (same as before)

### CODE REVIEW

File: pkg/server/config.go [hunk 1]
```diff
@@ -386,7 +386,7 @@ func MakeConfig(ctx context.Context, st *cluster.Settings) Config {
 		ScanInterval:                   defaultScanInterval,
 		ScanMaxIdleTime:                defaultScanMaxIdleTime,
 		EventLogEnabled:                defaultEventLogEnabled,
-		EnableWebSessionAuthentication: defaultEnableWebSessionAuthentication,
+		EnableWebSessionAuthentication: envutil.EnvOrDefaultBool("COCKROACH_REQUIRE_WEB_LOGIN", false),
```
Comment (mrtracy) [inline pkg/server/config.go]: Move this to a top-level var and reference it here.

**Review by mrtracy — COMMENTED** [review]
One thing to consider: there is an interaction with the "insecure" flag which is called out in the RFC for web session logins.  Once login is released, it will be *on* by default for secure servers, but disabled for insecure servers.  This is important because, in turn, it allows our frontend to detect whether or not authentication is enabled simply by checking the protocol (HTTP vs HTTPS). Otherwise, you need an additional mechanism to communicate this setting to the frontend so that it knows whether or not to require a login.

In other words: I think the system for enabling/disabling login is more complicated than is expressed here, i'm not sure we should be checking anything in until its fully considered.

**Comment by vilterp** [comment]
@mrtracy the env var only applies if we're in secure mode. Just updated the description to reflect that.

It's true that this is just to avoid myself, @couchand, and anyone else who is trying to test login having to twiddle a hardcoded boolean over the next couple months, so we can do that if necessary.

File: pkg/server/config.go (hunk 1 above)
Comment (vilterp) [reply pkg/server/config.go]: Done.

**Comment by mrtracy** [comment]
LGTM

File: pkg/server/config.go [hunk 2]
```diff
@@ -376,6 +375,8 @@ func MakeConfig(ctx context.Context, st *cluster.Settings) Config {
 		panic(err)
 	}
 
+	requireWebLogin := envutil.EnvOrDefaultBool("COCKROACH_REQUIRE_WEB_LOGIN", false)
```
Comment (couchand) [inline pkg/server/config.go]: If this is added it should definitely have `EXPERIMENTAL` in the name.

**Review by couchand — COMMENTED** [review]
As I mentioned to you yesterday, I'm not sure this is worth it.  It's doesn't seem to make things that much easier for us, and there's a risk involved.

Changing a line of code screams "this really isn't done yet".  Adding an env var (even one with EXPERIMENTAL in the name) seems like maybe it's okay to do.
